Add RefSlotHandle types to GC

rwy0717 opened the pull request:
The RefSlotHandle is "A handle to a reference-slot in an object". The
RefSlotHandle is a default implementation of a SlotHandle for bare,
unencoded gc-references.

This PR also adds a CompressedRefSlotHandle, for working with compressed
references. The reference shift is dynamic.

In the future, these classes will be used by OMR client's object
scanners to give the GC read-write capabilities on objects.

Signed-off-by: Robert Young <rwy0717@gmail.com>

charliegracie commented on gc/api/include/OMR/GC/CompressedRefSlotHandle.hpp:
```diff
@@ -0,0 +1,65 @@
+	void writeReference(omrobjectptr_t value) const { *_slot = encode(value, _shift); }
+
+private:
+	static omrobjectptr_t decode(uint32_t data, uint32_t shift)
```
Is it possible to modify the clang format for the GC such that this function would look like:
```
static omrobjectptr_t
decode(uint32_t data, uint32_t shift)
{
	return (omrobjectptr_t)(((uint64_t)data) << shift);
}
```

since that more closely matches the coding standards?

charliegracie replied:
Note that single line functions are ok just if it is a multi line function it should follow the above format.

rwy0717 replied:
You know, it's actually not as simple as I would have thought. Putting the type onto it's own line stops clang-format from putting short functions on one line. I'll just manually fix it up.

rwy0717 commented:
* Fixed formatting.
* Removed `#error` in normal refslot handle header, which was ensuring compressed-refs are disabled. It's not a valid test.

rwy0717 commented:
This is ready for a build

charliegracie commented:
@genie-omr build all

charliegracie commented:
Just to make sure is the `explicit` keyword available on all of our compilers?

rwy0717 commented:
I haven't actually tested it, but it's part of the original c++98 standard, so I assume so.

rwy0717 commented:
* Updated comments, copyright
* Rebased
* Removed explicit from CompressedRefSlotHandle, since it has no affect. The RefSlotHandle is still explicit.

rwy0717 commented:
This PR isn't needed anymore.
